sys/stdio_nimble/stdio_nimble.c

```diff
@@ -0,0 +1,399 @@
+
+            struct os_mbuf *om = ble_hs_mbuf_from_flat(_stdout_write_buf, len_read);
+
+            if (om != NULL) {
+                _sending_in_progress = true;
+                rc = ble_gattc_indicate_custom(_conn_handle, _val_handle_stdout, om);
+            }
+            else {
+                char str[80];
+                sprintf(str, "om == NULL; len_read = %d; _stdout_write_buf[MAX_MTU_PAYLOAD=%d]\n", len_read, MAX_MTU_PAYLOAD);
+                debug(str, strlen(str));
+            }
+        }
+    }
+
+    if (rc == BLE_HS_EAGAIN) {
+        d = "rc == BLE_HS_EAGAIN\n";debug(d, strlen(d));
+        /* try it again a few msec later */
+        ztimer_sleep(ZTIMER_MSEC, 5);
+        _send_from_ringbuffer(true);
+        return;
+    }
+    else if (rc == 0) {
+        d = "rc == 0\n";debug(d, strlen(d));
+        /* bytes were successfully sent, so remove them from the buffer */
+        ringbuffer_remove(&_rb_stdout, len_read);
+    }
+    else {
+        if (rc == -1) {
+            d = "rc = -1;\n"; debug(d, strlen(d));
+        }
+        else {
+            char str[40];
+            sprintf(str, "rc = %d; abort sending on error\n", rc);
+            debug(str, strlen(str));
+        }
+
+        /* abort sending on error */
+        _sending_in_progress = false;
+    }
+
+    mutex_unlock(&_send_mutex);
+}
+
+static void _clear_ringbuffer(void)
+{
+    ringbuffer_init(&_rb_stdin, _rb_stdin_mem, CONFIG_STDIO_NIMBLE_STDIN_BUFSIZE);
+
+    mutex_lock(&_rb_stdout_mutex);
+    ringbuffer_init(&_rb_stdout, _rb_stdout_mem, CONFIG_STDIO_NIMBLE_STDOUT_BUFSIZE);
+    mutex_unlock(&_rb_stdout_mutex);
+}
+
+static int _gap_event_cb(struct ble_gap_event *event, void *arg)
+{
+    (void) arg;
+    char* d;
+
+    switch (event->type) {
+
+        case BLE_GAP_EVENT_CONNECT:
+            if (event->connect.status == 0) {
+                _clear_ringbuffer();
+                _conn_handle = event->connect.conn_handle;
+                ble_att_set_preferred_mtu(MTU_MAX_NIMBLE);
+                ble_gattc_exchange_mtu(_conn_handle, NULL, NULL);
+            }
+            else {
+                _send_stdio = false;
+            }
+            break;
+
+        case BLE_GAP_EVENT_DISCONNECT:
+            _send_stdio = false;
+            break;
+
+        case BLE_GAP_EVENT_SUBSCRIBE:
+            if (event->subscribe.attr_handle == _val_handle_stdout) {
+                if (event->subscribe.cur_indicate == 1) {
+                    _send_stdio = true;
+                }
+                else {
+                    _send_stdio = false;
+                }
+            }
+            break;
+
+        case BLE_GAP_EVENT_NOTIFY_TX:
+            if (event->notify_tx.indication == 1) {
+                if (event->notify_tx.status == BLE_HS_EDONE) {
+                    d = "event->notify_tx.status == BLE_HS_EDONE\n"; debug(d, strlen(d));
+                    _send_from_ringbuffer(false);
+                }
```

Comment:
Doesn't this limit the rate? By waiting for BLE_HS_EDONE you prevent blocks of messages from being sent out together since its waiting for the ACK before proceeding with submitting the next payload batch when there might still be data to send, and since you don't re-send data, this does not seem needed no?

Reply:
With the new approach of callout functions and the state machine it is obsolete now